helix-view/src/document.rs

```diff
@@ -209,6 +322,127 @@ impl Document {
+                let is_last = idx == iter_len - 1;
+                let capacity = Self::calculate_decode_capacity(&mut decoder, chunk.as_bytes());
+                let mut total_read = 0;
+                buf.reserve(capacity.saturating_sub(buf.capacity()));
+                buf_changes.reserve(capacity.saturating_sub(buf_changes.capacity()));
+
+                // Loop until the entire chunk has been decoded into `buf`.
+                loop {
+                    let (result, read, ..) = decoder.decode_to_string(
+                        &chunk.as_bytes()[total_read..],
+                        &mut buf,
+                        is_last,
+                    );
+                    // Track how many bytes we have read so far, in case we need to allocate more
+                    // capacity to `buf`.
+                    total_read += read;
+
+                    // Check if we need to allocate more capacity to `buf`, otherwise append
+                    // to `changes`.
+                    match result {
+                        encoding_rs::CoderResult::InputEmpty => {
+                            debug_assert_eq!(total_read, chunk.len());
+                            let diff = similar::TextDiff::from_graphemes(chunk, &buf).grouped_ops(3);
+                            let diff_changes = diff.iter().flatten().filter_map(|op| {
+                                let (tag, chunk_range, buf_range) = op.as_tag_tuple();
+                                let start = chunk_range.start + total_rope_read;
+                                let end = chunk_range.end + total_rope_read;
+                                
+                                match tag {
+                                    similar::DiffTag::Insert | similar::DiffTag::Replace => {
+                                        let new_str = buf.tokenize_graphemes()[buf_range].concat();
+                                        Some((start, end, Some(new_str.into())))
+                                    }
+                                    similar::DiffTag::Delete => {
+                                        Some((start, end, None))
+                                    }
+                                    similar::DiffTag::Equal => None,
+                                }
+                            });
+                            buf_changes.extend(diff_changes);
+                            break;
+                        }
+                        encoding_rs::CoderResult::OutputFull => {
+                            debug_assert!(buf.len() > total_read);
+                            let needed_capacity = Self::calculate_decode_capacity(
+                                &mut decoder,
+                                &chunk.as_bytes()[total_read..],
+                            );
+                            buf.reserve(needed_capacity);
+                        }
+                    }
+                }
+                total_rope_read += chunk.chars().count();
```

Comment:
If there isn't a way to refactor the code to work in terms of bytes, then I would recommend using Ropey's `str_utils::byte_to_char_idx()` instead of `chars().count()`.  Specifically, this is the same as `chars().count()` except much faster:

```
byte_to_char_idx(chunk, chunk.len())
```

`byte_to_char_idx()` uses a bunch of SIMD and other cleverness to make counting utf8 chars much faster.